### PR TITLE
Switch metrics map to cilium/ebpf

### PR DESCRIPTION
**Pull request by jibi**
With this PR I'd like to start the conversation on migrating the logic to handle
eBPF maps to the cilium/ebpf library (and also get a sense of the effort
required for the switch).

This PR focuses on one specific map, the metrics map. We don't do anything too
complicated with it so I figured it would be a good first candidate.

The changes are split into 3 commits:

* the first one introduces a new ebpf package which wraps cilium/ebpf
  and re implements a minimal set of functionalities of the bpf package (the bare
  minimum required by the metricsmap package).
  My proposal would be to incrementally port features from the bpf package to
  the ebpf one on a map-by-map basis, but I wanted to hear your thoughts first
  regarding other strategies
* the second commit actually switches maps/metricsmap to use the ebpf
  package
* the third one gets rid of all the logic used to serialize and deserialize the
  metricsmap keys and values as to my understanding there's no need to do that
  (we can just access the actual map's keys and values).

Other questions:

* ebpf package design: should this new wrapper around cilium/ebpf expose roughly
  the same interface as the bpf package or are there shortcomings of the bpf
  package that should be addressed taking advantage of this rewrite? (Or
  alternatively we could do a first dumb rewrite and then once we are happy
  with the switch focus on the actual API of the ebpf package? Maybe that could
  help also with backports during the transition?)
* exporting cilium/ebpf: currently the ebpf package re-exports the
  cilium/ebpf identifiers that are required to use it (for example MapNew()
  expects an argument of type ebpf.MapSpec{}).
  I think in this way we could keep the ebpf package self contained (i.e. we
  don't require the user to import both ebpf and cilium/ebpf) at the cost of
  slightly increasing the complexity of the ebpf package as we need to manually
  reexport every identifier but I don't have strong opinions on that

### CODE REVIEW

**Comment by ti-mo** [comment]
Nice work, this is a great start for adopting the lib! I'd like to suggest an alternative to the callback-based approach taken here. An example is implemented here: https://github.com/mdlayher/netlink/blob/8014f9a7dbf4fd7b84a1783dd7b470db9113ff36/attribute.go#L144 (`AttributeDecoder`)

Instead of making the caller specify a callback, put the caller in charge of controlling the iterator (e.g. calling `Next()`). The iterator  pulls the key and the value into the respective `key, value` fields of the map and returns control to the caller. This has some advantages:

- If the caller is already aware of the contents of the map (or just wants bytes), the cost of `reflect` doesn't need to be paid on each step
- If the caller found the value they're looking for in this particular iteration, they can break the loop and yield the lock without being forced to scan all the way to the end
- If the caller is only interested in only keys or only values, we don't do any unnecessary processing

Of course, convenience wrappers can be added (like what `IterateWithCallback()` does), but the lower-level ops should be exposed for callers that don't need a full iteration.

**Comment by jibi** [comment]
> Nice work, this is a great start for adopting the lib! I'd like to suggest an alternative to the callback-based approach taken here. An example is implemented here: https://github.com/mdlayher/netlink/blob/8014f9a7dbf4fd7b84a1783dd7b470db9113ff36/attribute.go#L144 (`AttributeDecoder`)
> 
> Instead of making the caller specify a callback, put the caller in charge of controlling the iterator (e.g. calling `Next()`). The iterator pulls the key and the value into the respective `key, value` fields of the map and returns control to the caller. This has some advantages:
> 
>     * If the caller is already aware of the contents of the map (or just wants bytes), the cost of `reflect` doesn't need to be paid on each step
> 
>     * If the caller found the value they're looking for in this particular iteration, they can break the loop and yield the lock without being forced to scan all the way to the end
> 
>     * If the caller is only interested in only keys or only values, we don't do any unnecessary processing
> 
> 
> Of course, convenience wrappers can be added (like what `IterateWithCallback()` does), but the lower-level ops should be exposed for callers that don't need a full iteration.

I totally agree that lower-level ops should be exposed too :+1: 

For this specific case I went for the callback approach (even though I usually try to avoid them, mostly for readability reasons) since that map requires to be fully iterated every time anyway and also because that's the approach we are currently following by using `DumpWithCallback()`.

File: pkg/ebpf/map.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+	spec       *MapSpec
+	key, value interface{}
+
+	path string
+	lock lock.RWMutex
+}
+
+// NewMap creates a new Map object.
+func NewMap(spec *MapSpec, key, value interface{}) *Map {
```
Comment (brb) [inline pkg/ebpf/map.go]: IMO `key, value interface{}` is very unsafe. This allows any value to be written into a BPF map. Think of a potential bug which mixes values written into a map. Debugging such bugs in production could easily turn into a hell.

I'd suggest to introduce an interface type for key and value (same as we have in `pkg/bpf` today).

**Review by brb — COMMENTED** [review]
Nice! The iterative approach for adopting the library makes sense.

File: pkg/ebpf/map.go (hunk 1 above)
Comment (ti-mo) [reply pkg/ebpf/map.go]: These `key` and `value` interfaces are only used to derive a type from during `IterateWithCallback()`. The values specified here are never written to the map.

@brb your comment aligns with my suggestion of avoiding a callback-based model. IMO the iteration API should be limited to `Map.Next()` and then reading keys and values using `Map.{Key,Value}Uint16()`, `Map.{Key,Value}Bytes()` etc.

Comment (brb) [reply pkg/ebpf/map.go]: > These key and value interfaces are only used to derive a type from during IterateWithCallback(). The values specified here are never written to the map.

Yeah, but the returned map allows `(*cilium/ebpf.Map).Update(key, value interface{}, flags MapUpdateFlags) error` to be invoked which accepts any key/val.

> IMO the iteration API should be limited to Map.Next() and then reading keys and values using Map.{Key,Value}Uint16(), Map.{Key,Value}Bytes() etc.

I think it might be tricky to implement this correctly. See https://github.com/cilium/cilium/blob/master/pkg/bpf/map_linux.go#L665 for all hacks.

Comment (jibi) [reply pkg/ebpf/map.go]: > IMO key, value interface{} is very unsafe. This allows any value to be written into a BPF map. Think of a potential bug which mixes values written into a map. Debugging such bugs in production could easily turn into a hell.

That's a fair concern :+1: on the other hand I think the user of a map package like `maps/metricsmap` or `maps/lbmap` should not be concerned about this: the map package should handle (and hide) these details and expose only safe methods (ideally the user of `maps/*` should know nothing about the `ebpf` package). So once it's done and tested in the map package, there shouldn't be any risk of mixing different types

> I'd suggest to introduce an interface type for key and value (same as we have in pkg/bpf today).

I was trying to avoid that, if possible, as I think it introduces a good amount of complexity for little benefit (but I can only talk for the metricsmap case so far, I'm not sure if this will still hold when we start exposing the `Iterate()`/`Next()` methods)

> These key and value interfaces are only used to derive a type from during IterateWithCallback(). 

Good point. I think I'll move the logic to pass these key/value interfaces around directly in the `IterateWithCallback()` implementation of a given map package, to simplify the signature of `ebpf.NewMap()` (and also to save us a few calls to the `reflect` package)

**Comment by jibi** [comment]
test-me-please

File: pkg/ebpf/map.go (hunk 1 above)
Comment (ti-mo) [reply pkg/ebpf/map.go]: Just reading through it again, could we maybe do away with `IterateWithCallback()`? We should probably add something like `(*MapIterator).Dump()` and `(*MapIterator).AtomicDump()` (based on https://github.com/cilium/cilium/blob/master/pkg/bpf/map_linux.go#L665) to `cilium/ebpf` and create (or generate) wrappers that add typing in our internal subpackages (metricsmap, lbmap).

Just not a fan of the callback model in Go. :(

File: pkg/ebpf/map.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+// NewMap creates a new Map object.
+func NewMap(spec *MapSpec) *Map {
+	return &Map{
+		spec: spec,
+	}
+}
+
+// OpenOrCreate tries to open or create the eBPF map identified by the spec in
+// the Map object.
+func (m *Map) OpenOrCreate() error {
```
Comment (ti-mo) [inline pkg/ebpf/map.go:58]: This function does more than open/create, it pins it as well, which is not obvious from the fn name or docstring.

Since it doesn't take any arguments, wouldn't this code be better suited to live in `NewMap()` instead? Is there a requirement for this not to be tied to instantiation of the map?

For example, this can be added to `NewMap()` including a `pin bool` argument to control the pinning behaviour.

Comment (jibi) [reply pkg/ebpf/map.go:58]: > This function does more than open/create, it pins it as well which is not obvious from the fn name or docstring.

This function pins only if `MapSpec.Pinning` (passed to `ebpf.NewMap()`) is set to `ebpf.PinByName` (see for example https://github.com/cilium/cilium/blob/pr/jibi/move-metrics-to-cilium-ebpf/pkg/maps/metricsmap/metricsmap.go#L199)

> Since it doesn't take any arguments, wouldn't this code be better suited to live in NewMap() instead? Is there a requirement for this not to be tied to instantiation of the map?

I wanted to keep the current behavior as I'm not sure yet if there are occasions where it's actually useful to separate the map "definition" phase from the instantiation phase

**Comment by jibi** [comment]
test-me-please

**Comment by jibi** [comment]
test-gke

File: pkg/ebpf/map.go (hunk 1 above)
Comment (jibi) [reply pkg/ebpf/map.go]: I'm not sure I fully understand what you are proposing with `(*MapIterator).Dump()` and `(*MapIterator).AtomicDump()`, could you elaborate a bit more on that?

> Just not a fan of the callback model in Go. :(

I fully share the sentiment, but at the same time I don't think this should be a blocker until we hit a case that requires more control or atomicity over the iteration.

I think the current use case for metrics fits nicely with the `IterateWithCallback()` approach as (in my opinion):
* it doesn't make the code hard to follow:
https://github.com/cilium/cilium/blob/a096e381708d0585cb02aa331c06b06a05059d97/cilium/cmd/bpf_metrics_list.go#L72-L76 https://github.com/cilium/cilium/blob/a096e381708d0585cb02aa331c06b06a05059d97/pkg/maps/metricsmap/metricsmap.go#L187-L189
* it doesn't require any special handling (after all, we are just dumping metrics from a map which does never get an element removed)
* and also it's not sitting in any hot path

Comment (ti-mo) [reply pkg/ebpf/map.go]: > I think the current use case for metrics fits nicely with the `IterateWithCallback()` approach

I agree, let's port first and upstream changes later as we see common patterns emerge that would fit the external lib.

> I'm not sure I fully understand what you are proposing with `(*MapIterator).Dump()` and `(*MapIterator).AtomicDump()`

Oh, it just looks to me like some of this code could be simplified if `cilium/ebpf` map objects would have a method that returns their contents as a `map[interface{}]interface{}` (or `[]byte`s, ..) instead of the caller having to deal with iterators and having to write callbacks, etc..

**Comment by joestringer** [comment]
I didn't get a chance to review just yet, but I was wondering one thing - whether you have been able to do integration/runtime testing of the new version here? I'm not sure we have coverage of this. I'm guessing it's probably fine but if we suddenly broke the metrics CLI that would be suboptimal ;-)

**Comment by jibi** [comment]
> I didn't get a chance to review just yet, but I was wondering one thing - whether you have been able to do integration/runtime testing of the new version here? I'm not sure we have coverage of this. I'm guessing it's probably fine but if we suddenly broke the metrics CLI that would be suboptimal ;-)

There's a test case (for fragmented IPv4 packets) which uses [GetBPFPacketsCount](https://github.com/cilium/cilium/blob/master/test/helpers/wrappers.go#L203-L210) so at least indirectly we are testing it in the CI (and in addition to that I manually tested in the dev VM both the `bpf metrics list` command and the Prometheus exporter).

I agree that we should have better/proper coverage for this, I'll try to add something :+1:

**Comment by jibi** [comment]
test-me-please

File: pkg/ebpf/map.go (hunk 1 above)
Comment (brb) [reply pkg/ebpf/map.go]: > That's a fair concern +1 on the other hand I think the user of a map package like maps/metricsmap or maps/lbmap should not be concerned about this: the map package should handle (and hide) these details and expose only safe methods (ideally the user of maps/* should know nothing about the ebpf package). So once it's done and tested in the map package, there shouldn't be any risk of mixing different types

I think that a bug in a higher-level maps package is still possible. Also, out of laziness for a simple map it will be tempting not to introduce the higher-level package.

Comment (jibi) [reply pkg/ebpf/map.go]: Let me take a step back as I think this might have gone missed in the thread: I removed the `key, value interface{}` parameters from the `NewMap` method as we only need them in `IterateWithCallback()` (so I made that function responsible for handling the type conversions, which also saves us also a couple of call to the `reflect` package).

> I think that a bug in a higher-level maps package is still possible. 

Consider for example the `IterateWithCallback` method for `metricsmap`:

https://github.com/cilium/cilium/blob/7b780cc746785e09763723b2002e69113c1415d7/pkg/maps/metricsmap/metricsmap.go#L99-L106

whrere `IterateCallback` is defined as
https://github.com/cilium/cilium/blob/7b780cc746785e09763723b2002e69113c1415d7/pkg/maps/metricsmap/metricsmap.go#L33

I'm failing to see a way to have a mismatch for the `Key`/`Value` types as the callback is explicitly typed to receive a `*metricsmap.Key` and a `*metricsmap.Value` (and also how having a `bpf.Key` and `bpf.Value` interface would help catch these types of errors).

I think the same would hold for other methods that work with keys and values, for example I can imagine how the `Update()` method for the maglev map would look something like:

```go
func (m *MaglevInnerMap) Update(key *MaglevInnerKey, val *MaglevInnerVal) error {
    return m.Update(key, val, 0);
}
```

so again, a small layer of abstraction could prevent types mismatch.

> Also, out of laziness for a simple map it will be tempting not to introduce the higher-level package.

I really don't think we should design this taking laziness into account. Having a per-map higher level package gives us:

* type safety as the user is forced to use the correct keys and values types (as opposed to using directly `cilium/ebpf`, where all functions which work with maps expect `interface{}`s)
* less code duplication
* a consistent interface across the different map packages

**Comment by jibi** [comment]
test-me-please

**Comment by jibi** [comment]
test-me-please

File: pkg/ebpf/ebpf.go [hunk 3]
```diff
@@ -0,0 +1,24 @@
+package ebpf
+
+// Copyright 2021 Authors of Cilium
```
Comment (christarazi) [inline pkg/ebpf/ebpf.go:1]: Nit: I think the copyright passage should be starting from the first line

File: pkg/ebpf/map.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+type IterateCallback func(key, value interface{})
+
+// Map represents an eBPF map.
+type Map struct {
+	*ciliumebpf.Map
+
+	spec *MapSpec
+
+	path string
+	lock lock.RWMutex
```
Comment (christarazi) [inline pkg/ebpf/map.go]: Nit: can you move the mutex above the fields that it protects?

File: pkg/ebpf/map_register.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+	return mapRegister[name]
+}
+
+// GetOpenMaps returns a slice of all open BPF maps. This is identical to
+// calling GetMap() on all open maps.
+func GetOpenMaps() []*models.BPFMap {
+	// create a copy of mapRegister so we can unlock the mutex again as
+	// locking Map.lock inside of the mutex is not permitted
+	mutex.RLock()
+	maps := []*Map{}
```
Comment (christarazi) [inline pkg/ebpf/map_register.go]:
```
maps := make([]*Map, 0, len(mapRegister))
```

File: cilium/cmd/bpf_metrics_list.go [hunk 6]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2018-2021 Authors of Cilium
+// Copyright 2018 Authors of Cilium
```
Comment (christarazi) [inline cilium/cmd/bpf_metrics_list.go]: This should be reversed?

**Review by christarazi — COMMENTED** [review]
Mostly minor nits

**Comment by jibi** [comment]
test-me-please

**Comment by jibi** [comment]
retest-gke

**Comment by jibi** [comment]
retest-gke

**Comment by jibi** [comment]
retest-netnext

**Comment by jibi** [comment]
test-me-please

**Comment by jibi** [comment]
retest-netnext

**Comment by jibi** [comment]
retest-netnext

File: pkg/ebpf/map.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+type Map struct {
+	lock lock.RWMutex
+	*ciliumebpf.Map
+
+	spec *MapSpec
+	path string
+}
+
+// NewMap creates a new Map object.
+func NewMap(spec *MapSpec) *Map {
```
Comment (brb) [inline pkg/ebpf/map.go:50]: Why can't we use `ciliumebpf.MapSec` directly?

File: pkg/ebpf/map.go (hunk 1 above)
Comment (brb) [reply pkg/ebpf/map.go]: Thanks for the explanation. Any idea how we can prevent "direct" usage of underlying `cilium/ebpf` maps without the small layer of abstraction?

File: pkg/ebpf/map.go (hunk 7 above)
Comment (jibi) [reply pkg/ebpf/map.go:50]: I'm trying keep the `ebpf` package self contained as:
* ideally the user of this package should not care about the fact that it's backed by `ciliumebpf` (it's just an implementation detail)
* naming clashing: having to import `ebpf` and `ciliumebpf` could generate some confusion so having just a single package would keep everything simpler

at the same time I don't feel super strongly about that, so that's just an attempt for now (which so far is working as I only needed to reexport a couple of identifiers but we may have to revisit this as we migrate more maps)

File: pkg/ebpf/map.go (hunk 1 above)
Comment (jibi) [reply pkg/ebpf/map.go]: > Any idea how we can prevent "direct" usage of underlying cilium/ebpf maps without the small layer of abstraction?

I don't have any suggestion to actually enforce that beside adding some documentation on the package and code review :disappointed:

Comment (joestringer) [reply pkg/ebpf/map.go]: `[ "$(grep github.com/cilium/ebpf pkg/ | grep -v pkg/ebpf | wc -l)" != "0" ]` ?

**Review by brb — APPROVED** [review]
:rocket:

File: pkg/ebpf/map.go (hunk 1 above)
Comment (jibi) [reply pkg/ebpf/map.go]: Nice! That would work :boom: (although we couldn't enable this right now as there are a few packages which are already using `cilium/ebpf` and need to be updated).

Should this check live in `contrib/scripts`, so be part of the Go prechecks?

**Review by twpayne — APPROVED** [review]
LVGTM

**Comment by jibi** [comment]
test-me-please

**Comment by jibi** [comment]
test-me-please

File: pkg/ebpf/map.go [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+// GetModel returns a BPF map in the representation served via the API.
+func (m *Map) GetModel() *models.BPFMap {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+
+	mapModel := &models.BPFMap{
+		Path: m.path,
+	}
+
+	// TODO: handle map cache. See pkg/bpf/map_linux.go:GetModel()
```
Comment (aanm) [inline pkg/ebpf/map.go:111]: When is this todo going to be addressed?

File: pkg/maps/metricsmap/metricsmap.go [hunk 9]
```diff
@@ -139,16 +123,31 @@ func (k *Key) DropForwardReason() string {
...
+	c := uint64(0)
+	for _, v := range *vs {
+		c += v.Count
+	}
+
+	return c
+}
+
+// Bytes() returns the sum of all the per-CPU bytes values
+func (vs *Values) Bytes() uint64 {
```
Comment (aanm) [inline pkg/maps/metricsmap/metricsmap.go]:
```golang
// Bytes returns the sum of all the per-CPU bytes values
func (vs Values) Bytes() uint64 {
	b := uint64(0)
	for _, v := range vs {
```

File: pkg/maps/metricsmap/metricsmap.go [hunk 10]
```diff
@@ -139,16 +123,31 @@ func (k *Key) DropForwardReason() string {
...
-
 // IsDrop checks if the reason is drop or not.
 func (k *Key) IsDrop() bool {
 	return k.Reason == monitorAPI.DropInvalid || k.Reason >= monitorAPI.DropMin
 }
 
+// Count() returns the sum of all the per-CPU count values
+func (vs *Values) Count() uint64 {
+	c := uint64(0)
+	for _, v := range *vs {
```
Comment (aanm) [inline pkg/maps/metricsmap/metricsmap.go]:
```golang
// Count returns the sum of all the per-CPU count values
func (vs Values) Count() uint64 {
	c := uint64(0)
	for _, v := range vs {
```

**Review by aanm — COMMENTED** [review]
Some non-blocking nits.

File: pkg/ebpf/map.go (hunk 8 above)
Comment (jibi) [reply pkg/ebpf/map.go:111]: My idea was to implement the caching logic when we migrate to the ebpf package a map that actually uses this feature (so that it can be properly tested)

**Comment by jibi** [comment]
test-me-please

**Review by aanm — APPROVED** [review]

**Comment by jibi** [comment]
retest-gke
